core/src/main/java/org/elasticsearch/cluster/routing/allocation/UnassignedShardDecision.java

```diff
@@ -0,0 +1,188 @@
+import org.elasticsearch.common.Nullable;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Represents the allocation decision by an allocator for an unassigned shard.
+ */
+public class UnassignedShardDecision {
+    /** a constant representing a shard decision where no decision was taken */
+    public static final UnassignedShardDecision DECISION_NOT_TAKEN =
+        new UnassignedShardDecision(false, null, null, null, null, null, null);
+
+    private final boolean decisionTaken;
+    @Nullable
+    private final Decision decision;
+    @Nullable
+    private final AllocationStatus allocationStatus;
+    @Nullable
+    private final String explanation;
+    @Nullable
+    private final String assignedNodeId;
+    @Nullable
+    private final String allocationId;
+    @Nullable
+    private final Map<String, Decision> nodeDecisions;
+
+    private UnassignedShardDecision(boolean decisionTaken,
+                                    Decision decision,
+                                    AllocationStatus allocationStatus,
+                                    String explanation,
+                                    String assignedNodeId,
+                                    String allocationId,
+                                    Map<String, Decision> nodeDecisions) {
+        assert decision != null || decisionTaken == false : "if a decision was taken, we must have the accompanying decision";
+        assert explanation != null || decisionTaken == false : "if a decision was taken, there must be an explanation for it";
+        assert assignedNodeId != null || decisionTaken == false || decision.type() != Type.YES :
+            "a yes decision must have a node to assign the shard to";
+        assert allocationStatus != null || decisionTaken == false || decision.type() == Type.YES :
+            "only a yes decision should not have an allocation status";
+        assert allocationId == null || assignedNodeId != null : "allocation id can only be null if the assigned node is null";
+        this.decisionTaken = decisionTaken;
+        this.decision = decision;
+        this.allocationStatus = allocationStatus;
+        this.explanation = explanation;
+        this.assignedNodeId = assignedNodeId;
+        this.allocationId = allocationId;
+        this.nodeDecisions = nodeDecisions != null ? Collections.unmodifiableMap(nodeDecisions) : null;
+    }
+
+    /**
+     * Creates a NO decision with the given {@link AllocationStatus} and explanation for the NO decision.
+     */
+    public static UnassignedShardDecision noDecision(AllocationStatus allocationStatus, String explanation) {
+        Objects.requireNonNull(explanation, "explanation must not be null");
+        Objects.requireNonNull(allocationStatus, "allocationStatus must not be null");
+        return new UnassignedShardDecision(true, Decision.NO, allocationStatus, explanation, null, null, null);
```

Comment:
I think maybe this should call `noDecision(allocationStatus, explanation, null)` so it goes through a single call chain and can be changed in less places in the future?